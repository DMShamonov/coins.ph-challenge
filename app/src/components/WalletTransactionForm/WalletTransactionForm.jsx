// @flow

import * as React from 'react';

import {
  Form,
  FormTitle,
  FormGroup,
  FormError,
} from 'components/Form';
import Input from 'components/Input';
import Button from 'components/Button';

import type { PropsType } from './types.js.flow';

export default function WalletTransactionForm({
  address,
  amount,
  walletPassphrase,
  error,
  hasError,
  isFormDisabled,
  onAddressChange,
  onAmountChange,
  onWalletPassphraseChange,
  onSubmit,
}: PropsType): React.Node {
  return (
    <Form onSubmit={onSubmit}>
      <FormTitle>Create transaction</FormTitle>
      <FormGroup>
        <Input
          name="address"
          label="Address"
          placeholder="Enter a wallet address"
          value={address}
          required
          disabled={isFormDisabled}
          onChange={onAddressChange}
        />
      </FormGroup>
      <FormGroup>
        <Input
          type="number"
          name="amount"
          label="Amount"
          placeholder="Enter an amount of satoshi"
          value={amount || ''}
          required
          disabled={isFormDisabled}
          onChange={onAmountChange}
        />
      </FormGroup>
      <FormGroup>
        <Input
          type="password"
          name="walletPassphrase"
          label="Wallet passphrase"
          placeholder="Enter a wallet passphrase"
          value={walletPassphrase}
          autoComplete="off"
          required
          disabled={isFormDisabled}
          onChange={onWalletPassphraseChange}
        />
      </FormGroup>
      {hasError && <FormError>{error}</FormError>}
      <Button type="submit" disabled={isFormDisabled}>Create transaction</Button>
    </Form>
  );
}
