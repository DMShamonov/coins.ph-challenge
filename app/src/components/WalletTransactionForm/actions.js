import { batch } from 'react-redux';

import api, { OTP } from 'controllers/api';

import WALLET_TRANSACTION_FORM_ACTION_TYPES from 'redux/actionTypes/walletTransactionForm';
import WALLET_TRANSACTIONS_ACTION_TYPES from 'redux/actionTypes/walletTransactions';
import MODAL_ACTION_TYPES from 'redux/actionTypes/modal';
import { getErrorMessage } from 'controllers/utilities';

// eslint-disable-next-line import/prefer-default-export
export function createWalletTransaction(walletId) {
  return async function _createWallet(dispatch, getState) {
    dispatch({ type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_START_PROCESSING });

    const {
      ui: { walletTransactionForm: { address, amount, walletPassphrase } },
    } = getState();

    try {
      await api().unlock({ otp: OTP });

      const wallet = await api().wallets().get({ id: walletId });
      const { hash } = await wallet.sendCoins({ address, amount, walletPassphrase });
      const transaction = await wallet.getTransaction({ id: hash });

      batch(() => {
        dispatch({
          type: WALLET_TRANSACTIONS_ACTION_TYPES.ON_ADD,
          payload: { transactions: [transaction] },
        });
        dispatch({ type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_RESET_FORM });
        dispatch({ type: MODAL_ACTION_TYPES.ON_HIDE });
      });
    } catch (e) {
      const error = getErrorMessage(e) || 'Something went wrong. Try again, please.';

      dispatch({ type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_SET_ERROR, error });
    }

    dispatch({ type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_FINISH_PROCESSING });
  };
}
