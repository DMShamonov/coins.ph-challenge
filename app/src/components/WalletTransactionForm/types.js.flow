// @flow

export type PropsType = {|
  address: string,
  amount: number,
  walletPassphrase: string,
  error: string,
  hasError: boolean,
  isFormDisabled: boolean,
  onAddressChange: SyntheticInputEvent<HTMLInputElement> => void,
  onAmountChange: SyntheticInputEvent<HTMLInputElement> => void,
  onWalletPassphraseChange: SyntheticInputEvent<HTMLInputElement> => void,
  onSubmit: Event => void,
|};
