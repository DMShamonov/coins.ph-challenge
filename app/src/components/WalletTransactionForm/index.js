import { connect } from 'react-redux';

import WALLET_TRANSACTION_FORM_ACTION_TYPES from 'redux/actionTypes/walletTransactionForm';

import { createWalletTransaction } from './actions';
import WalletTransactionForm from './WalletTransactionForm';

function mapStateToProps({ ui: { walletTransactionForm } }) {
  const { isProcessing, error, ...formData } = walletTransactionForm;
  const hasError = !!error;

  return {
    ...formData,
    error,
    hasError,
    isFormDisabled: isProcessing,
  };
}

function mapDispatchToProps(dispatch, { walletId }) {
  return {
    onAddressChange({ target: { value: address } }) {
      dispatch({
        type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_CHANGE_ADDRESS,
        payload: { address },
      });
    },
    onAmountChange({ target: { value } }) {
      const amount = parseFloat(value);

      dispatch({
        type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_CHANGE_AMOUNT,
        payload: { amount: amount > 0 ? amount : 0 },
      });
    },
    onWalletPassphraseChange({ target: { value: walletPassphrase } }) {
      dispatch({
        type: WALLET_TRANSACTION_FORM_ACTION_TYPES.ON_CHANGE_WALLET_PASSPHRASE,
        payload: { walletPassphrase },
      });
    },
    onSubmit(e) {
      e.preventDefault();

      dispatch(createWalletTransaction(walletId));
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(WalletTransactionForm);
