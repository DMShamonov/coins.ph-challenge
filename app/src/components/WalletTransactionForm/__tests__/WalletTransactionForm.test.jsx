import React from 'react';
import { shallow } from 'enzyme';

import WalletTransactionForm from '../WalletTransactionForm';

describe('`WalletTransactionForm` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow((
      <WalletTransactionForm
        address="__ADDRESS__"
        amount={100000000}
        walletPassphrase="__WALLET_PASSPHRASE__"
        error=""
        hasError={false}
        isFormDisabled={false}
        onAddressChange={jest.fn()}
        onAmountChange={jest.fn()}
        onWalletPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `error` prop is set', () => {
    expect(shallow((
      <WalletTransactionForm
        address="__ADDRESS__"
        amount={100000000}
        walletPassphrase="__WALLET_PASSPHRASE__"
        error="__ERROR__"
        hasError={false}
        isFormDisabled={false}
        onAddressChange={jest.fn()}
        onAmountChange={jest.fn()}
        onWalletPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `hasError` prop is `true`', () => {
    expect(shallow((
      <WalletTransactionForm
        address="__ADDRESS__"
        amount={100000000}
        walletPassphrase="__WALLET_PASSPHRASE__"
        error=""
        hasError
        isFormDisabled={false}
        onAddressChange={jest.fn()}
        onAmountChange={jest.fn()}
        onWalletPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `error` prop is set and `hasError` prop is `true`', () => {
    expect(shallow((
      <WalletTransactionForm
        address="__ADDRESS__"
        amount={100000000}
        walletPassphrase="__WALLET_PASSPHRASE__"
        error="__ERROR__"
        hasError
        isFormDisabled={false}
        onAddressChange={jest.fn()}
        onAmountChange={jest.fn()}
        onWalletPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `isFormDisabled` prop is `true`', () => {
    expect(shallow((
      <WalletTransactionForm
        address="__ADDRESS__"
        amount={100000000}
        walletPassphrase="__WALLET_PASSPHRASE__"
        error=""
        hasError={false}
        isFormDisabled
        onAddressChange={jest.fn()}
        onAmountChange={jest.fn()}
        onWalletPassphraseChange={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });
});
