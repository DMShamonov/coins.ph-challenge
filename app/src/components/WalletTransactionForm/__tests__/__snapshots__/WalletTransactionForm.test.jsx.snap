// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`WalletTransactionForm\` component should render correctly when \`error\` prop is set 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create transaction
  </FormTitle>
  <FormGroup>
    <Input
      disabled={false}
      label="Address"
      name="address"
      onChange={[MockFunction]}
      placeholder="Enter a wallet address"
      required={true}
      value="__ADDRESS__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      disabled={false}
      label="Amount"
      name="amount"
      onChange={[MockFunction]}
      placeholder="Enter an amount of satoshi"
      required={true}
      type="number"
      value={100000000}
    />
  </FormGroup>
  <FormGroup>
    <Input
      autoComplete="off"
      disabled={false}
      label="Wallet passphrase"
      name="walletPassphrase"
      onChange={[MockFunction]}
      placeholder="Enter a wallet passphrase"
      required={true}
      type="password"
      value="__WALLET_PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={false}
    type="submit"
  >
    Create transaction
  </Button>
</Form>
`;

exports[`\`WalletTransactionForm\` component should render correctly when \`error\` prop is set and \`hasError\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create transaction
  </FormTitle>
  <FormGroup>
    <Input
      disabled={false}
      label="Address"
      name="address"
      onChange={[MockFunction]}
      placeholder="Enter a wallet address"
      required={true}
      value="__ADDRESS__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      disabled={false}
      label="Amount"
      name="amount"
      onChange={[MockFunction]}
      placeholder="Enter an amount of satoshi"
      required={true}
      type="number"
      value={100000000}
    />
  </FormGroup>
  <FormGroup>
    <Input
      autoComplete="off"
      disabled={false}
      label="Wallet passphrase"
      name="walletPassphrase"
      onChange={[MockFunction]}
      placeholder="Enter a wallet passphrase"
      required={true}
      type="password"
      value="__WALLET_PASSPHRASE__"
    />
  </FormGroup>
  <FormError>
    __ERROR__
  </FormError>
  <Button
    disabled={false}
    type="submit"
  >
    Create transaction
  </Button>
</Form>
`;

exports[`\`WalletTransactionForm\` component should render correctly when \`hasError\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create transaction
  </FormTitle>
  <FormGroup>
    <Input
      disabled={false}
      label="Address"
      name="address"
      onChange={[MockFunction]}
      placeholder="Enter a wallet address"
      required={true}
      value="__ADDRESS__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      disabled={false}
      label="Amount"
      name="amount"
      onChange={[MockFunction]}
      placeholder="Enter an amount of satoshi"
      required={true}
      type="number"
      value={100000000}
    />
  </FormGroup>
  <FormGroup>
    <Input
      autoComplete="off"
      disabled={false}
      label="Wallet passphrase"
      name="walletPassphrase"
      onChange={[MockFunction]}
      placeholder="Enter a wallet passphrase"
      required={true}
      type="password"
      value="__WALLET_PASSPHRASE__"
    />
  </FormGroup>
  <FormError />
  <Button
    disabled={false}
    type="submit"
  >
    Create transaction
  </Button>
</Form>
`;

exports[`\`WalletTransactionForm\` component should render correctly when \`isFormDisabled\` prop is \`true\` 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create transaction
  </FormTitle>
  <FormGroup>
    <Input
      disabled={true}
      label="Address"
      name="address"
      onChange={[MockFunction]}
      placeholder="Enter a wallet address"
      required={true}
      value="__ADDRESS__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      disabled={true}
      label="Amount"
      name="amount"
      onChange={[MockFunction]}
      placeholder="Enter an amount of satoshi"
      required={true}
      type="number"
      value={100000000}
    />
  </FormGroup>
  <FormGroup>
    <Input
      autoComplete="off"
      disabled={true}
      label="Wallet passphrase"
      name="walletPassphrase"
      onChange={[MockFunction]}
      placeholder="Enter a wallet passphrase"
      required={true}
      type="password"
      value="__WALLET_PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={true}
    type="submit"
  >
    Create transaction
  </Button>
</Form>
`;

exports[`\`WalletTransactionForm\` component should render correctly with default props 1`] = `
<Form
  onSubmit={[MockFunction]}
>
  <FormTitle>
    Create transaction
  </FormTitle>
  <FormGroup>
    <Input
      disabled={false}
      label="Address"
      name="address"
      onChange={[MockFunction]}
      placeholder="Enter a wallet address"
      required={true}
      value="__ADDRESS__"
    />
  </FormGroup>
  <FormGroup>
    <Input
      disabled={false}
      label="Amount"
      name="amount"
      onChange={[MockFunction]}
      placeholder="Enter an amount of satoshi"
      required={true}
      type="number"
      value={100000000}
    />
  </FormGroup>
  <FormGroup>
    <Input
      autoComplete="off"
      disabled={false}
      label="Wallet passphrase"
      name="walletPassphrase"
      onChange={[MockFunction]}
      placeholder="Enter a wallet passphrase"
      required={true}
      type="password"
      value="__WALLET_PASSPHRASE__"
    />
  </FormGroup>
  <Button
    disabled={false}
    type="submit"
  >
    Create transaction
  </Button>
</Form>
`;
